### PR TITLE
peering: expose servers over Kubernetes service

### PR DESCRIPTION
**Pull request by ndhanushkodi**
Changes proposed in this PR:
- Add new expose-servers service to Helm chart
- Support scraping addresses from expose-servers service when type LB or Nodeport and using those to generate the token
- Bump consul/api module to use generate token endpoint changes from https://github.com/hashicorp/consul/pull/13844
- Update acceptance tests to deploy 3 servers, and on clouds it'll use a load balancer service, on kind itll use a nodeport service. 

How I've tested this PR:
- ran acceptance tests on Kind, GKE, EKS, AKS
- also tested Nodeport manually on GKE with 3 nodes because we can only run with 1 server on Kind.

Requires the consul peering dialer to be tweaked to retry more quickly when dialing a follower instead of leader. For now this is running off of a branch of consul, but eventually will use changes like: https://github.com/hashicorp/consul/pull/13938  

How I expect reviewers to test this PR:
You could set up two clusters, deploy with peering enabled, and servers enabled which will cause the expose service to get deployed. Then you could apply a dialer and acceptor, and inspect the token, and see that it has the LB or nodeport addresses. You can also test  a service dialing another service.

Checklist:
- [x] Tests added
- [ ] CHANGELOG entry added 
  > HashiCorp engineers only, community PRs should not add a changelog entry.
  > Entries should use present tense (e.g. Add support for...)

### CODE REVIEW

File: acceptance/tests/peering/peering_connect_namespaces_test.go
Comment (thisisnotashwin) [inline acceptance/tests/peering/peering_connect_namespaces_test.go:69]: might require a little uncommenting :P

Comment (ndhanushkodi) [reply acceptance/tests/peering/peering_connect_namespaces_test.go]: yup!

**Comment by ndhanushkodi** [comment]
@thisisnotashwin I addressed your comments from the now closed PR here so re-requesting your review :) tqsm!!

File: acceptance/tests/peering/peering_connect_test.go
Comment (thisisnotashwin) [inline acceptance/tests/peering/peering_connect_test.go:140]: same comment about vanilla `retry.Run`

File: charts/consul/values.yaml
Comment (thisisnotashwin) [inline charts/consul/values.yaml:1085]: between now and GA of this feature, @ishustava, you and I and the team should have a conversation about potentially swapping out the client to use 8503 and the servers to use 8502 for GRPC given agentless should ensure no competition for port, in which case I would like the helm default to be the same as the Consul defaults. I have kinda led us down the 8503 track and would like to get us to a feasible long term solution.

**Review by thisisnotashwin — APPROVED** [review]
I have a few suggestion but im good with it getting merged once they are addressed..don't think I need to review this again! 😄 
excellent work here Nitya!! you crushed it!!

File: .circleci/config.yml
Comment (ndhanushkodi) [reply .circleci/config.yml:105]: I'll revert all the changes to this file before merging! I'll only keep the change that removes the -disable-peering flag from AKS

File: .circleci/config.yml
Comment (ndhanushkodi) [reply .circleci/config.yml:138]: same as above, will be reverted before merging!

File: charts/consul/values.yaml
Comment (ndhanushkodi) [reply charts/consul/values.yaml:1085]: That sounds like a good plan!

File: acceptance/tests/peering/peering_connect_namespaces_test.go
Comment (ndhanushkodi) [reply acceptance/tests/peering/peering_connect_namespaces_test.go:185]: This step could potentially take longer now, because the controller will try to read the service address, and error if for example, a load balancer IP doesn't yet exist. So it could take a couple loops through the controller to get a real address here, so a retry of longer than 7s could be very possible, so I increased the timer

File: acceptance/tests/peering/peering_connect_test.go
Comment (ndhanushkodi) [reply acceptance/tests/peering/peering_connect_test.go:140]: Yup! replied above

File: acceptance/tests/peering/peering_connect_namespaces_test.go
Comment (thisisnotashwin) [reply acceptance/tests/peering/peering_connect_namespaces_test.go:185]: aha!! thanks for the clarification!!

File: charts/consul/templates/expose-servers-service.yaml
Comment (ishustava) [inline charts/consul/templates/expose-servers-service.yaml:54]: Could we make the service port `8502` so that it's like consul's default?

File: control-plane/connect-inject/peering_acceptor_controller.go
Comment (ishustava) [inline control-plane/connect-inject/peering_acceptor_controller.go:465]: we should get the port from the k8s service instead of hard-coding it. It's not blocking since it matches how we define the service for now, but I think it'll be more future-proof if we get it from there.

**Review by ishustava — APPROVED** [review]
Great work, Nitya! This is very thorough!! I'm leaving some comments, but they are mostly non-blocking, so will not hold up the PR on those.

File: acceptance/framework/k8s/deploy.go
Comment (ndhanushkodi) [reply acceptance/framework/k8s/deploy.go:99]: The leader routine to set up the peering stream can take longer now with the Load balancer in between, so it was helpful to increase this timeout to allow the peering stream more time to be set up, so the connection can be successful.

File: acceptance/tests/peering/peering_connect_namespaces_test.go
Comment (ndhanushkodi) [reply acceptance/tests/peering/peering_connect_namespaces_test.go:141]: nice thanks!

File: acceptance/tests/peering/peering_connect_namespaces_test.go
Comment (ndhanushkodi) [reply acceptance/tests/peering/peering_connect_namespaces_test.go:158]: nice thanks!

File: charts/consul/test/unit/expose-servers-service.bats
Comment (ndhanushkodi) [reply charts/consul/test/unit/expose-servers-service.bats:12]: Yup, nice catch!

File: control-plane/connect-inject/peering_acceptor_controller.go
Comment (ndhanushkodi) [reply control-plane/connect-inject/peering_acceptor_controller.go:439]: ah nice

File: acceptance/framework/k8s/deploy.go
Comment (ishustava) [reply acceptance/framework/k8s/deploy.go:99]: Ahh got it!

File: charts/consul/test/unit/expose-servers-service.bats
Comment (ndhanushkodi) [reply charts/consul/test/unit/expose-servers-service.bats:157]: added now!

File: control-plane/connect-inject/peering_acceptor_controller.go
Comment (ndhanushkodi) [reply control-plane/connect-inject/peering_acceptor_controller.go]: going to do this one later!

File: charts/consul/templates/expose-servers-service.yaml
Comment (ndhanushkodi) [reply charts/consul/templates/expose-servers-service.yaml]: yup updated

**Pull request merged**
